src/par_iter/enumerate.rs

```diff
@@ -82,38 +74,32 @@ impl<M> IndexedParallelIterator for Enumerate<M>
 ///////////////////////////////////////////////////////////////////////////
 // Producer implementation
 
-pub struct EnumerateProducer<'p, P>
-    where P: Producer<'p>,
-{
+pub struct EnumerateProducer<P> {
     base: P,
     offset: usize,
-    phantoms: PhantomType<&'p ()>,
 }
 
-impl<'p, P> Producer<'p> for EnumerateProducer<'p, P>
-    where P: Producer<'p>
+impl<P> Producer for EnumerateProducer<P>
+    where P: Producer
 {
-    type Item = (usize, P::Item);
-    type Shared = P::Shared;
-
-    fn cost(&mut self, shared: &Self::Shared, items: usize) -> f64 {
-        self.base.cost(shared, items) // enumerating is basically free
+    fn cost(&mut self, items: usize) -> f64 {
+        self.base.cost(items) // enumerating is basically free
     }
 
-    unsafe fn split_at(self, index: usize) -> (Self, Self) {
+    fn split_at(self, index: usize) -> (Self, Self) {
         let (left, right) = self.base.split_at(index);
         (EnumerateProducer { base: left,
-                             offset: self.offset,
-                             phantoms: PhantomType::new() },
+                             offset: self.offset },
          EnumerateProducer { base: right,
-                             offset: self.offset + index,
-                             phantoms: PhantomType::new() })
+                             offset: self.offset + index })
     }
+}
+
+impl<P> IntoIterator for EnumerateProducer<P> where P: Producer {
+    type Item = (usize, P::Item);
+    type IntoIter = iter::Enumerate<P::IntoIter>;
 
-    unsafe fn produce(&mut self, shared: &Self::Shared) -> (usize, P::Item) {
-        let item = self.base.produce(shared);
-        let index = self.offset;
-        self.offset += 1;
-        (index, item)
+    fn into_iter(self) -> Self::IntoIter {
+        self.base.into_iter().enumerate()
```

Comment:
The current tests are clearly not enough in this case. :) 

For the most part I haven't stressed about testing them in parallel because the producers themselves were essentially stateless, and the type system ensures that we don't screw up too badly, but I think some targeted tests would make a lot of sense. (ie., manually invoking the relevant methods)